# CS3237 Lab 2 - Neural Networks


# CS3237 Lab 2 - Neural Networks (MARKING SCHEME)

|Student Nunmber|Name                  |
|:--------------|:---------------------|
|               | Enter your name here |
|               | Enter your name here |

Please work together as a team of 2 to complete this lab. You will need to submit ONE copy of this notebook per team, but please fill in the names of both team members above. This lab is worth 25 marks:

**DO NOT SUBMIT MORE THAN ONE COPY OF THIS LAB!**

## SUBMISSION INSTRUCTIONS

Please submit this completed Jupyter Notebook (cs3237lab2.ipynb) to the Files -> Labs -> Lab Submissions -> Lab2 Submissions-> Group Bxx folder by 1.00 PM on the following dates:

 1. Group B1: Friday 3 September, 1 pm.
 2. Group B2: Monday 6 September, 1 pm.
 3. Group B3: Sunday 5 September, 1 pm.
 
## LATE SUBMISSION POLICY

If you submit between 1 pm and 1.15 pm, a 5 mark penalty will be levied. Submission is strictly not allowed once the folder closes, and you will receive 0 for the lab. NO EXCEPTIONS WILL BE MADE.

## 1. Introduction

The objectives of this lab are:

    1. To familiarize you with how to create dense neural networks using Keras.
    2. To familiarize you with how to encode input and output vectors for neural networks.
    3. To give you some insight into how hyperparameters like learning rate and momentum affect training.
    
To save time we will train each experiment only for 50 epochs. This will lead to less than optimal results but is enough for you to make observations.

**HINT: YOU CAN HIT SHIFT-ENTER TO RUN EACH CELL. NOTE THAT IF A CELL IS DEPENDENT ON A PREVIOUS CELL, YOU WILL NEED TO RUN THE PREVIOUS CELL(S) FIRST **


## 2. The Irises Dataset

We will now work again on the Irises Dataset, which we used in Lab 1, for classifying iris flowers into one of three possible types. As before we will consider four factors:

    1. Sepal length in cm
    2. Sepal width in cm
    3. Petal length in cm
    4. Petal width in cm

In this dataset there are 150 sample points. The code below loads the dataset and prints the first 10 rows so we have an idea of what it looks like.

In [87]:
from sklearn.datasets import load_iris
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

iris = load_iris()

print("First 10 rows of data:")
print(iris.data[:10])

First 10 rows of data:
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]


### 2.2 Scaling the Data

We make use of the MinMaxScaler to scale the inputs to between 0 and 1.  The code below does this and prints the first 10 rows again, to show us the difference.

In the next section we will investigate what happens if we use unscaled data.

In [88]:
scaler = MinMaxScaler()
scaler.fit(iris.data)
X = scaler.transform(iris.data)

print("First 10 rows of SCALED data.")
print(X[:10])

First 10 rows of SCALED data.
[[0.22222222 0.625      0.06779661 0.04166667]
 [0.16666667 0.41666667 0.06779661 0.04166667]
 [0.11111111 0.5        0.05084746 0.04166667]
 [0.08333333 0.45833333 0.08474576 0.04166667]
 [0.19444444 0.66666667 0.06779661 0.04166667]
 [0.30555556 0.79166667 0.11864407 0.125     ]
 [0.08333333 0.58333333 0.06779661 0.08333333]
 [0.19444444 0.58333333 0.08474576 0.04166667]
 [0.02777778 0.375      0.06779661 0.04166667]
 [0.16666667 0.45833333 0.08474576 0.        ]]


### 2.3 Encoding the Targets

In Lab 1 we saw that the target values (type of iris flower) is a vector from 0 to 2. We can see the 150 labels below:


In [89]:
print(iris.target)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


We can use this to train the neural network, but we will use "one-hot" encoding, where we have a vector of _n_ integers consisting of 0's and 1's.  The table below shows how one-hot encoding works:

|   Value    |    One-Hot Encoding    |
|:----------:|:----------------------:|
| 0 | \[1 0 0\] |
| 1 | \[0 1 0\] |
| 2 | \[0 0 1\] |

Keras provides the to_categorical function to create one-hot vectors:



In [90]:
from tensorflow.keras.utils import to_categorical

Y = to_categorical(y = iris.target, num_classes = 3)
print(Y)

[[1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0.

Now let's split the data into training and testing data:



In [91]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, 
                                                    random_state = 1)


### 2.4 Building our Neural Network

Let's now begin building a simple neural network with a single hidden layer, using the Stochastic Gradient Descent (SGD) optimizer, ReLu transfer functions for the hidden layer and softmax for the output layer.

The code to do this is shown below:

In [92]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD

# Create the neural network
nn = Sequential()
nn.add(Dense(100, input_shape = (4, ), activation = 'relu'))
nn.add(Dense(3, activation = 'softmax'))

# Create our optimizer
sgd = SGD(learning_rate = 0.1, momentum = 0.01, nesterov = False)

# 'Compile' the network to associate it with a loss function,
# an optimizer, and what metrics we want to track
nn.compile(loss='categorical_crossentropy', optimizer=sgd, 
          metrics = 'accuracy')

### 2.5 Training the Neural Network

As is usually the case, we can call the "fit" method to train the neural network for 50 epochs. We will shuffle the training data between epochs, and provide validation data.

In [93]:
nn.fit(X_train, Y_train, shuffle = True, epochs = 50, 
      validation_data = (X_test, Y_test))

Epoch 1/50
4/4 [==============================] - 0s 37ms/step - loss: 1.0751 - accuracy: 0.3000 - val_loss: 1.0631 - val_accuracy: 0.2000
Epoch 2/50
4/4 [==============================] - 0s 7ms/step - loss: 1.0214 - accuracy: 0.3917 - val_loss: 1.0320 - val_accuracy: 0.2667
Epoch 3/50
4/4 [==============================] - 0s 7ms/step - loss: 0.9763 - accuracy: 0.5417 - val_loss: 1.0005 - val_accuracy: 0.4667
Epoch 4/50
4/4 [==============================] - 0s 7ms/step - loss: 0.9360 - accuracy: 0.6750 - val_loss: 0.9727 - val_accuracy: 0.5667
Epoch 5/50
4/4 [==============================] - 0s 7ms/step - loss: 0.8974 - accuracy: 0.6917 - val_loss: 0.9369 - val_accuracy: 0.5667
Epoch 6/50
4/4 [==============================] - 0s 6ms/step - loss: 0.8630 - accuracy: 0.6917 - val_loss: 0.9054 - val_accuracy: 0.5667
Epoch 7/50
4/4 [==============================] - 0s 7ms/step - loss: 0.8291 - accuracy: 0.6917 - val_loss: 0.8661 - val_accuracy: 0.5667
Epoch 8/50
4/4 [=================

---
#### Question 1

Run the code above. Do you see evidence of underfitting? Overfitting? Justify your answers. ***(4 MARKS)***

**Answer: No. Both the training and validation accuracy are high and are similar in value.**

_(For TA) Marks awarded: ____ / 4_

---

#### Question 2a

Consult the documentation for the SGD optimizer [here](https://keras.io/api/optimizers/sgd/). What does the lr parameter do? ***(1 MARK)***

**Answer: The lr parameter is the learning rate, a hyperparameter that controls how much to change the model in response to the estimated error each time the model weights are updated.**

#### Question 2b

The documentation states that the momentum parameter "accelerates gradient descent in the relevant direction and dampens oscillations". Using Google or other means, illustrate what this means. ***(2 MARKS)***

**Answer: Momentum is an extension to the gradient descent optimization algorithm that allows the search to build inertia in a direction in the search space and overcome the oscillations of noisy gradients and coast across flat spots of the search space. It decreases the number of function evaluations required to reach the optima, or to improve the capability of the optimization algorithm, thus leading to better final results. **

_(For TA) Marks awarded: ____ / 3_

----

#### Question 3a

We will now play with the lr parameter. Adjust the lr parameter to the following values and record the final training and validation accuracies in the respective columns. Also observe the sequence of accuracies over the training period, and place your observation in the "remarks" column, e.g. "Progresses steadily", "some oscillation" etc. ***(3 MARKS)***

**Answer: Fill the table below **

|  lr    | Training Acc. | Validation Acc. |      Remarks      |
|:------:|---------------|-----------------|-------------------|
|0.01    | 0.6917        | 0.5667          | Peaks early, then settles with low accuracy                       |
|0.1     | 0.9500        | 0.9333          | Generally steady progression, with some oscillation               |
|1.0     | 0.9583        | 1.0000          | Generally steady progression, with some oscillation               |
|10.0    | 0.3750        | 0.3667          | Accuracies fluctuate without any improvement                      |
|100     | 0.3083        | 0.3667          | Accuracies fluctuate without any improvement                      |
|1000    | 0.2417        | 0.2000          | Accuracies fluctuate without any improvement                      |
|10000   | 0.2833        | 0.4333          | Accuracies fluctuate without any improvement                      |
|100000  | 0.3083        | 0.4333          | Accuracies fluctuate without any improvement                      |


#### Question 3b

Based on your observations above, comment on the effect of small and very large learning rates on the learning. ***(2 MARKS)***

**Answer: A learning rate that is too small may not achieve a good enough accuracy, but a learning rate that is too large can result in tthe model converging too quickly, delivering a suboptimal solution. **

_(For TA) Marks awarded: ____ / 5_

### 2.5 Using Momentum

We will now experiment with the momentum term. To do this:

    1. Change the learning rate to 0.1.
    2. Set the momentum to 0.1. Note: Do not use the Nesterov parameter - Leave it as False.
    
Run your neural network.

---

#### Question 4a

Keeping the learning rate at 0.1, complete the table below using the momentum values shown. Again record any observations in the "Remarks" column. ***(3 MARKS)***

**Answer: Fill the table below**

| momentum | Training Acc. | Validation Acc. |      Remarks      |
|:--------:|---------------|-----------------|-------------------|
|0.001     |0.9417         |0.9667           |                   |
|0.01      |0.9333         |0.9667           |                   |
|0.1       |0.9250         |0.9333           |                   |
|1.0       |0.9000         |0.9667           |                   |

#### Question 4b

Based on your observations above, does the momentum term help in learning? ***(2 MARKS)***

**Answer: Momentum does not have much impact on learning. The training and validation accuracies are high and close in value despite the change in momentum value. **

_(For TA) Marks awarded: ____ / 5_

---

### 2.6 Using Raw Unscaled Data

We begin by using unscaled X and Y data. The code below will create 120 training samples and 30 testing samples (20% of the total of 150 samples):

X_unscaled = iris.data
Y_raw = iris.target
X_utrain, X_utest, Y_utrain, Y_utest = train_test_split(X_unscaled, Y,
                                                        test_size = 0.2,
                                                        random_state = 1)

---

#### Question 5

Create a new neural network called "nn2" below using a single hidden layer of 100 neurons. Train using the data in X_utrain, X_utest and validate with Y_utrain and Y_utest. Again use the SGD optimizer with a learning rate of 0.1 and no momentum, and train for 50 epochs. ***(3 marks)***

In [96]:
"""
Enter your code for Question 5 below. To TA: 3 marks for code.
"""
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD

# Create the neural network
nn2 = Sequential()
nn2.add(Dense(100, input_shape = (4, ), activation = 'relu'))
nn2.add(Dense(3, activation = 'softmax'))

# Create our optimizer
sgd2 = SGD(learning_rate = 0.1)

# 'Compile' the network to associate it with a loss function,
# an optimizer, and what metrics we want to track
nn2.compile(loss='categorical_crossentropy', optimizer=sgd2, 
          metrics = 'accuracy')

# Train nn2
nn2.fit(X_utrain, Y_utrain, shuffle = True, epochs = 50, 
      validation_data = (X_utest, Y_utest))

Epoch 1/50
4/4 [==============================] - 0s 40ms/step - loss: 1.7182 - accuracy: 0.3667 - val_loss: 1.1018 - val_accuracy: 0.5333
Epoch 2/50
4/4 [==============================] - 0s 7ms/step - loss: 0.8088 - accuracy: 0.6250 - val_loss: 0.6733 - val_accuracy: 0.8000
Epoch 3/50
4/4 [==============================] - 0s 8ms/step - loss: 0.6402 - accuracy: 0.7583 - val_loss: 0.6153 - val_accuracy: 0.5667
Epoch 4/50
4/4 [==============================] - 0s 7ms/step - loss: 0.5218 - accuracy: 0.7583 - val_loss: 0.4857 - val_accuracy: 0.8000
Epoch 5/50
4/4 [==============================] - 0s 7ms/step - loss: 0.5573 - accuracy: 0.6750 - val_loss: 0.6167 - val_accuracy: 0.5667
Epoch 6/50
4/4 [==============================] - 0s 8ms/step - loss: 0.5058 - accuracy: 0.7083 - val_loss: 0.4597 - val_accuracy: 0.7333
Epoch 7/50
4/4 [==============================] - 0s 7ms/step - loss: 0.4471 - accuracy: 0.7333 - val_loss: 0.3987 - val_accuracy: 1.0000
Epoch 8/50
4/4 [=================

**(Question 5 continues)**

Observe the training and validation error. Does not scaling the input affect the training? Why do you think this is so? What is the advantage of scaling? ***(5 MARKS)***

**Answer: From the results observed, there is minimal impact of scaling on the dataset. If a dataset is not scaled, the feature with a larger value range will dominate when calculating distances. In this dataset, the features refer to the sepal/petal lengths and widths, and the values are rather close to each other. Scaling thus has minimal impact on this dataset. Significant bias will be observed in features that have a bigger range of values, and will be more impactful on the neural network's learning. Scaling will hence be advatageous in this case, as it will remove differences in magnitutde and achieve faster convergence, hence helping the model learn better. **

_(For TA) Marks awarded: ____ / 8_

---

## 3 Conclusion

In this lab we saw how to create a simple Dense neural network to complete the relatively simple task of learning how to classify irises according to their sepal and petal characteristics. 


---

***FOR TA ONLY***

| Question |  Marks  |
|:--------:|:-------:|
|1         |     /4  |
|2         |     /3  |
|3         |     /5  |
|4         |     /5  |
|5         |     /8  |
|Total:    |     /25 |

